core/src/rest/http_client.ts

```diff
@@ -0,0 +1,240 @@
+  tokenType?: TokenType;
+  token?: string;
+  userAgent: string;
+  baseURL: string;
+  version: number;
+  maxRetries = 5;
+
+  private queue = new Map<string, Queue>();
+
+  constructor(options?: HTTPClientOptions) {
+    if (options !== undefined) {
+      if (
+        (options.token !== undefined && options.tokenType === undefined) ||
+        (options.token === undefined && options.tokenType !== undefined)
+      ) {
+        throw new TypeError(
+          `Either tokenType and token both must be specified or neither of should be.`,
+        );
+      }
+    }
+    this.tokenType = options?.tokenType;
+    this.token = options?.token;
+    this.userAgent = options?.userAgent ?? USER_AGENT;
+    this.baseURL = options?.baseURL ?? DISCORD_API_BASE;
+    this.version = options?.version ?? DISCORD_API_VERSION;
+    this.maxRetries = options?.maxRetries ?? this.maxRetries;
+  }
+
+  async request(
+    method: HTTPMethod,
+    endpoint: Endpoint,
+    options?: RequestOptions,
+  ) {
+    const bucket = options?.bucket ?? getRouteBucket(endpoint);
+    const queue = this.queue.get(bucket) ??
+      (this.queue.set(bucket, new Queue()), this.queue.get(bucket)!);
+
+    const execute = async (): Promise<any> => {
+      await queue.wait();
+
+      // Enqueue it and wait for the queue to be resolved.
+      try {
+        const headers: Record<string, string> = {
+          "user-agent": this.userAgent,
+        };
+
+        if (this.token !== undefined && this.tokenType !== undefined) {
+          headers["authorization"] = `${this.tokenType} ${this.token}`.trim();
+        }
+
+        if (options?.headers !== undefined) {
+          Object.assign(headers, options.headers);
+        }
+
+        let body = undefined;
+        if (
+          ["POST", "PATCH", "PUT"].includes(method)
+        ) {
+          // It's multipart formdata.
+          if (options?.body instanceof FormData) {
+            body = options.body;
+          } else {
+            if (typeof options?.body === "object") {
+              // It's a simple JSON body.
+              body = JSON.stringify(
+                options.body,
+                (_, v) => typeof v === "bigint" ? v.toString() : v,
+              );
+              headers["content-type"] = "application/json";
+            } else {
+              // Default to empty body to avoid Cloudflare errors.
+              body = "";
+            }
+          }
+        }
+
+        const res = await fetch(
+          `${this.baseURL}/v${this.version}${endpoint}`,
+          {
+            method,
+            headers,
+            body,
+          },
+        );
+
+        const resetTime = Date.now() +
+          Number(res.headers.get("X-RateLimit-Reset-After")) * 1000 + 5; // add a little bit of delay to avoid 429s
+        queue.resetTime = resetTime;
+        if (res.headers.get("X-RateLimit-Global") === "true") {
+          this.queue.forEach((q) => {
+            if (q.resetTime < resetTime) {
```

Comment:
wouldn't this always be true?
because you first need to wait the queue time before you can go.
So whatever happens between the first wait and this will always be lower than the `resetTime`